### PR TITLE
Soret effect for cylindrical and spherical systems

### LINKED ISSUE #818: Soret effect in cylindrical/spherical coordinates
Currently, the Soret effect can be used with cartesian meshes. It would be nice to extend the functionality for cylindrical and spherical cases.

### PR DESCRIPTION
**Pull request by KulaginVladimir**
## Proposed changes

This PR fixes #818 and allows using the Soret effect with cylindrical/spherical meshes. To verify the implementation, MMS tests were added, but for steady-state cases only. 

I also removed the previous error message and the related test.

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
One thing to keep in mind is that I don't thing the non-Cartesian derived quantity for Surface Flux support Soret?

Also, I'd be curious to see verification of this implementation not with a Dirichlet BC but a flux BC

**Comment by KulaginVladimir** [comment]
> Also, I'd be curious to see verification of this implementation not with a Dirichlet BC but a flux BC

You mean a second pair of MMS tests?

**Comment by RemDelaporteMathurin** [comment]
Yes. Since the flux changes with Soret I'd like to verify that a flux bc is correctly set

File: test/system/test_cylindrical_coordinates.py [hunk 1]
```diff
@@ -112,3 +112,107 @@ def test_temperature_MMS():
...
+        Args:
+            u (sympy.Matrix): a sympy vector field
+
+        Returns:
+            sympy.Expr: the divergence of u
+        """
+        return sp.simplify(sp.diff(r * u[0], r) / r + sp.diff(u[1], z))
+
+    # Create and mark the mesh
+    fenics_mesh = fenics.UnitSquareMesh(250, 250)
```
[the remark this answers is not in the document]
Comment (KulaginVladimir) [reply test/system/test_cylindrical_coordinates.py:146]: if a lower L2 error is ok, then yes. For 100x100 mesh, it'd be 2.5e-4.

**Comment by KulaginVladimir** [comment]
Maybe, we don't need the tests on $r\in[0,1]$? We can define a problem on $r\in[1,2]$ and set `DirichletBC` on the left surface and set `FluxBC` on the left right.

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 2]
```diff
@@ -95,7 +96,8 @@ def test_h_flux_with_soret(self):
 @pytest.mark.parametrize("height", [2, 3])
 @pytest.mark.parametrize("c_top", [2, 3])
 @pytest.mark.parametrize("c_bottom", [2, 3])
-def test_compute_cylindrical(r0, radius, height, c_top, c_bottom):
+@pytest.mark.parametrize("Soret", [False, True])
+def test_compute_cylindrical(r0, radius, height, c_top, c_bottom, Soret):
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:100]: Let's follow the [PEP8 recommendations](https://peps.python.org/pep-0008/#function-and-variable-names:~:text=Function%20names%20should,retain%20backwards%20compatibility.) here and use lowercase

```suggestion
@pytest.mark.parametrize("soret", [False, True])
def test_compute_cylindrical(r0, radius, height, c_top, c_bottom, soret):
```

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 3]
```diff
@@ -105,6 +107,7 @@ def test_compute_cylindrical(r0, radius, height, c_top, c_bottom):
         height (float): cylinder height
         c_top (float): concentration top
         c_bottom (float): concentration bottom
+        Soret (bool): the Soret flag
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:110]:
```suggestion
        soret (bool): if True, Soret effect will be set
```

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 4]
```diff
@@ -155,14 +158,36 @@ def test_compute_cylindrical(r0, radius, height, c_top, c_bottom):
         / height
         * (0.5 * r1**2 - 0.5 * r0**2)
     )
+
+    if Soret:
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:162]:
```suggestion
    if soret:
```

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 5]
```diff
@@ -155,14 +158,36 @@ def test_compute_cylindrical(r0, radius, height, c_top, c_bottom):
...
+            * float(my_flux.D)
+            * float(my_flux.Q)
+            * c_fun(z1)
+            / k_B
+            / T(z1) ** 2
+            * growth_rate
+            * (0.5 * r1**2 - 0.5 * r0**2)
+        )
+
+    computed_value = my_flux.compute(soret=Soret)
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:180]:
```suggestion
    computed_value = my_flux.compute(soret=soret)
```

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 6]
```diff
@@ -155,14 +158,36 @@ def test_compute_cylindrical(r0, radius, height, c_top, c_bottom):
...
     assert np.isclose(computed_value, expected_value)
 
 
 @pytest.mark.parametrize("r0", [1, 2])
 @pytest.mark.parametrize("radius", [1, 2])
 @pytest.mark.parametrize("c_left", [3, 4])
 @pytest.mark.parametrize("c_right", [1, 2])
-def test_compute_spherical(r0, radius, c_left, c_right):
+@pytest.mark.parametrize("Soret", [False, True])
+def test_compute_spherical(r0, radius, c_left, c_right, Soret):
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:190]:
```suggestion
@pytest.mark.parametrize("soret", [False, True])
def test_compute_spherical(r0, radius, c_left, c_right, soret):
```

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 7]
```diff
@@ -201,19 +226,40 @@ def test_compute_spherical(r0, radius, c_left, c_right):
...
 
     my_flux.n = f.FacetNormal(mesh_fenics)
     my_flux.ds = ds
 
-    computed_value = my_flux.compute()
     # expected value is the integral of the flux over the surface
     flux_value = float(my_flux.D) * (c_left - c_right) / (r1 - r0)
     expected_value = -4 * math.pi * flux_value * r1**2
+
+    if Soret:
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:243]:
```suggestion
    if soret:
```

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py [hunk 8]
```diff
@@ -201,19 +226,40 @@ def test_compute_spherical(r0, radius, c_left, c_right):
...
+            * float(my_flux.D)
+            * float(my_flux.Q)
+            * c_fun(r1)
+            / k_B
+            / T(r1) ** 2
+            * growth_rate
+            * r1**2
+        )
+
+    computed_value = my_flux.compute(soret=Soret)
```
Comment (RemDelaporteMathurin) [inline test/unit/test_exports/test_derived_quantities/test_surface_flux.py:261]:
```suggestion
    computed_value = my_flux.compute(soret=soret)
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
All looks good to me! Thanks for the thorough testing.

I would just use lowercase for variable names. Uppercase is usually reserved for class names.

See the [PEP8 guide](https://peps.python.org/pep-0008/#function-and-variable-names)

**Comment by RemDelaporteMathurin** [comment]
The CI is failing due to an import issue. This is not a FESTIM issue but maybe a change in the conda recipe

File: test/unit/test_exports/test_derived_quantities/test_surface_flux.py (hunk 3 above)
Comment (KulaginVladimir) [reply test/unit/test_exports/test_derived_quantities/test_surface_flux.py]: I included your suggestions. Also, added this line to the docstrings of the spherical case.

**Review by RemDelaporteMathurin — APPROVED** [review]

**Pull request merged**